keps/sig-node/127-user-namespaces/README.md

```diff
@@ -0,0 +1,456 @@
+host files).
+
+#### Story 5
+
+As a cluster admin, I want to use different host UIDs/GIDs for pods running in
+the same node (whenever kernel/kube features allow it), so I can mitigate the
+impact a compromised pod can have on other pods and the node itself.
+
+### Notes/Constraints/Caveats
+
+#### UID space
+
+#### Volume support
+
+### Risks and Mitigations
+
+## Design Details
+
+Note: Names are preliminary yet, I'm using field names to simplify explanations.
+
+### Pod.spec changes
+
+The following changes will be done to the pod.spec:
+
+- `pod.spec.useHostUsers`: bool.
+If true or not present, uses the host user namespace (as today)
+If false, a new userns is created for the pod.
+This field will be used for phase 1, 2 and 3.
+
+- `pod.spec.securityContext.userns.pod2podIsolation`: Enum
+If enabled, we will make the userns mappings be non-overlapping as much as possible.
+This field will be used in phase 3.
+
+### Phases
+
+We propose to divide the work in 3 phases. Each phase makes this work with
+either more isolation or more workloads. When no support is yet added to handle
+some workload, a clear error will be shown.
+
+Please note the last sub-section here is a table with the summary of the changes
+proposed on each phase.
+
+#### Phase 1: pods "without" volumes
+
+This phase makes pods "without" volume work with user namespaces. This is
+activated via the bool `pod.spec.HostUsers` and works only when no volume or any
+combination of the following volume types are used by the pod:
+
+ - configmap
+ - secret
+ - downwardAPI
+ - emptyDir
+ - projected
+
+This list of volumes was chosen as they can't be used to share files with other
+pods.
+
+The mapping length will be 65535, mapping the range 0-65534 to the pod. This wide
+range makes sure most workloads will work fine.
+
+The mapping will be chosen by the kubelet, using a simple algorithm to give
+different pods in this category ("without" volumes) a non-overlapping mapping.
+Giving non-overlapping mappings generates the best isolation for pods.
+
+Furthermore, the node UID space of 2^32 can hold up to 2^16 pods each with a
+mapping length of 65535 (2^16-1) top. This imposes a limit of 65k pods per node,
+but that is not an issue we will hit in practice for a long time, if ever (today
+we run 110 pods per node by default).
+
+#### Phase 2: pods with volumes
+
+This phase makes user namespaces work in pods with volumes too. This is
+activated via the bool `pod.spec.HostUsers` too and pods fall into this mode if
+some other volume type than the ones listed for phase 1 is used. IOW, when phase
+2 is implemented, pods that use volume types not supported in phase 1, fall into
+the phase 2.
+
+All pods in this mode will use _the same_ mapping, chosen by the kubelet, with a
+length 65535, and mapping the range 0-65534 too. IOW, each pod will have its own user
+namespace, but they will map to _the same_ UIDs/GIDs in the host.
```

Comment:
We need to ensure that this mapping is the same for all nodes, right?  In other words we HAVE TO spec the exact mapping (and never use that  for non-volume-using pods).  We should just go full spec:

e.g. in the host userNS, assuming allocations of 64K

[0]: reserved for host
[65535]: reserved for host
[1]: reserved for pods using "Cluster" userNS
[2]-[4098]: reserved for per-pod userNSes
[4098]-[65535]: reserved for later aallocation

Reply:
@AkihiroSuda do you think a static allocation would be a problem for rootless?  In theory, running in rootless mode still allows to create user namespaces, that are children of the rootless user namespace.

If we define only:

 [off:0, length:65536]: reserved for host
 [off:65536, length:65536]: reserved for pods using "Cluster" userNS
 [off:2*65536, length:AS_MANY_IDS_LEFT]: reserved for per-pod userNS

would it work more easily for rootless?

Reply:
I'm not sure I understand the concern.  The reason I left a lot of room reserved is because (I am 97% sure) we need a way to get deterministic mappings for pods that DO use volumes.  Even if we don't tackle it in phase 1 or 2, we need it, so let's leave room until we know more.

Concretely:

Start a pod in namespace N using a PVC.  The first time it starts up, it will get some mapping of inside UID (let's call it U1) to outside UID (U2).  It will write files. to that PVC as U1 which get written to FS as U2, right?  Then that pod will get moved to another node.  When it starts up, it is still U1 inside, and the files are still U2 outside.  So the same mapping is needed. 

We discussed this in the previous draft.  I'm OK to not do it in phase 1, but let's not make it impossible to do.  If we find a better way, then we can expand the per-pod userNS range at that time.

Reply:
My suggestion was about moving the reserved area before the per-pod userNS.

In this way when running in rootless mode, if there are not enough IDs allocated to the user, only the unique number of per-pod will be affected.

If we use Linux idmapped mounts, there is no requirement on keeping the same mapping on different nodes though.

We can mount the volume with an idmap that maps U1 in the user namespace to U1 on the host, another advantage with this model is that pods with different mappings can access the same volume.

An example of how it works at runtime (used Podman for simplicity since it already supports idmapped mounts):

```
# stat -c %u:%g /volume/file
0:0

# podman run -v /volume:/volume:idmap --uidmap 0:2000:1000 --gidmap 0:5000:1000 --rm alpine sh -c 'echo uids; cat /proc/self/uid_map; echo gids; cat /proc/self/gid_map; stat -c %u:%g /volume/file'
uids
         0       2000       1000
gids
         0       5000       1000
0:0
```
The command above specifies the mapping for the user namespace and with the `idmap` option we request to use the container user namespace for the volume idmap.

What happens is that both the container and the host see the same mappings, and the conversion happens at runtime in the kernel.

A pod with a different user namespace mapping can use the same volume without any conflict.

Reply:
Maybe I gravely misunderstand idmapped mounts.  I went off and re-read the docs, but I am confidused by your description.

Assume that I have 2 containers running, both of which use a volume (e.g. NFS or some other shared-writable FS).  How do I ensure that both containers have access to the filesystem?   It seems they need to be the same kernel-UID?  That says to me that we need some sort of global-scope allocator or some set of ranges set aside for deterministically derivable mapping (e.g. hash of NS name).

What am I missing?

Reply:
@thockin you are missing what I already explained in the other thread when you said this ;) https://github.com/kubernetes/enhancements/pull/2101#issuecomment-954663014

ID mapped mounts DO solve the issue of having containers with different mapping sharing volumes. They do an id shifting (one on read and the reverse on write) so the UID written to the volumes is the same as the user running inside the container. I explained this in more detail in that link. ID mapped mounts solve this problem (although just for fs that support ID mapped mounts, that is the big caveat), it was developed by Christian Brauner to solve this issue with containers in LXC/LXD (besides other use cases too, systemd is using this, etc.).

Reply:
The ID shifting was explained in more detail in last years Linux Plumbers conf, a talk by Christian. I can try to search for it if you want more details. Or you can play locally if you have a recent kernel, Christian did a simple user-space application (linked in his github, but also the cover letter of the patch series), that was useful for me to play locally.

Also, runc and CRI-o (not the runtime-spec yet, sadly) added support for id mapped mounts. There are no trivial issues to solve to support this, that is why I'd like to get it out of the way at least for phase 1 :)

Reply:
Agree to not solve in phase 1 or even 2.  My only ask is "leave room for unknown-unknowns" - don't allocate the whole host space just yet :)

Reply:
Added to the list of unresolved stuff. Not because I want to challenge it (at least not now, I'll see later), just so we don't forget to incorporate or challenge this :D